x-pack/auditbeat/module/system/package/rpm_linux.go

```diff
@@ -113,27 +111,21 @@ my_rpmtsFree(void *f, rpmts ts) {
 // By default, librpm is going to trap various UNIX signals including SIGINT and SIGTERM
 // which will prevent Beats from shutting down correctly.
 //
-// This disables that behavior. We should be very dilligent in
+// This disables that behavior by nullifying rpmsqEnable. We should be very dilligent in
 // cleaning up in our use of librpm.
 //
 // More recent versions of librpm have a new function rpmsqSetInterruptSafety()
 // to do this, see below.
 //
 // See also:
 // - librpm traps signals and calls exit(1) to terminate the whole process incl. our Go code: https://github.com/rpm-software-management/rpm/blob/rpm-4.11.3-release/lib/rpmdb.c#L640
-// - has caused problems for gdb before, calling rpmsqEnable(_, NULL) is the workaround they also use: https://bugzilla.redhat.com/show_bug.cgi?id=643031
+// - has caused problems for gdb before, they also nullify rpmsqEnable: https://bugzilla.redhat.com/show_bug.cgi?id=643031
 // - the new rpmsqSetInterruptSafety(), unfortunately only available in librpm>=4.14.0 (CentOS 7 has 4.11.3): https://github.com/rpm-software-management/rpm/commit/56f49d7f5af7c1c8a3eb478431356195adbfdd25
-void
-my_disableLibrpmSignalTraps(void *f) {
-	int (*rpmsqEnable)(int, rpmsqAction_t);
-	rpmsqEnable = (int (*)(int, rpmsqAction_t))f;
-
-	// Disable all traps
-	rpmsqEnable(-SIGHUP, NULL);
-	rpmsqEnable(-SIGINT, NULL);
-	rpmsqEnable(-SIGTERM, NULL);
-	rpmsqEnable(-SIGQUIT, NULL);
-	rpmsqEnable(-SIGPIPE, NULL);
+extern int rpmsqEnable (int signum, void *handler);
+int
+rpmsqEnable (int signum, void *handler)
```

Comment:
I wonder who is calling this function. Is it called internally by librpm?

If that's the case I'm surprised that you can override it.

Reply:
Yes, it's internal to librpm. This works because we dynamically load librpm using `dlopen()` after the program is started. Any symbols that are already defined at that point will not be redefined.